### ISSUE TITLE
error: use of overloaded operator '&' is ambiguous

### ISSUE THREAD

**Issue by duffy-corelight** [zeek/spicy#219]
```
module SS7LayersPcapOrNg;
function reverseOrderDigits(onePast: uint8, thisOne: uint8) {
}

function label_then_BCD(contentExceptLast: bytes, lastByte: uint8) : string {
  for (idx in contentExceptLast) {
    reverseOrderDigits(idx, uint8(0xF) & (idx >> 4));
  }
  return "%s" % (lastByte,);
}
```
reports
```
3380 /usr/local/etc/tmp$spicy/build/bin/spicyc -j Issue_219_bitwise_ambiguous.spicy
SS7LayersPcapOrNg.cc:21:78: error: use of overloaded operator '&' is ambiguous (with operand types 'hilti::rt::integer::safe<std::uint8_t>' (aka 'SafeInt<unsigned char, detail::SafeIntException>') and 'SafeInt<char, hilti::rt::integer::detail::SafeIntException>')
        reverseOrderDigits(idx, (hilti::rt::integer::safe<std::uint8_t>{15u} & idx >> hilti::rt::integer::safe<std::uint64_t>{4u}));
                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/etc/tmp/spicy/hilti/include/hilti/3rdparty/SafeInt/SafeInt.hpp:6222:34: note: candidate function [with U = SafeInt<char, hilti::rt::integer::detail::SafeIntException>]
    _CONSTEXPR14 SafeInt< T, E > operator &( U rhs ) const SAFEINT_NOTHROW
                                 ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/3rdparty/SafeInt/SafeInt.hpp:7080:30: note: candidate function [with T = char, U = SafeInt<unsigned char, hilti::rt::integer::detail::SafeIntException>, E = hilti::rt::integer::detail::SafeIntException]
_CONSTEXPR11 SafeInt< T, E > operator &( U lhs, SafeInt< T, E > rhs ) SAFEINT_NOTHROW
                             ^
/usr/local/etc/tmp/spicy/hilti/include/hilti/3rdparty/SafeInt/SafeInt.hpp:6216:34: note: candidate function
    _CONSTEXPR14 SafeInt< T, E > operator &( SafeInt< T, E > rhs ) const SAFEINT_NOTHROW
                                 ^
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, int)
        reverseOrderDigits(idx, (hilti::rt::integer::safe<std::uint8_t>{15u} & idx >> hilti::rt::integer::safe<std::uint64_t>{4u}));
                                                                             ^
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(int, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(long long, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(__int128, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned int, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned long long, unsigned __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, __int128)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, unsigned int)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, unsigned long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, unsigned long long)
SS7LayersPcapOrNg.cc:21:78: note: built-in candidate operator&(unsigned __int128, unsigned __int128)
1 error generated.
[error] jit: failed to execute compilation action.
[error] jit: failed to compile C++ code unit SS7LayersPcapOrNg to bitcode
[error] spicyc: JIT compilation failed
```

**Comment by rsmmr** [comment]
This might be the same as #216.

**Comment by bbannier** [comment]
> This might be the same as #216.

I think this is another issue. For above example we generate
```
static auto __hlt::SS7LayersPcapOrNg::label_then_BCD(const hilti::rt::Bytes& contentExceptLast, const hilti::rt::integer::safe<uint8_t>& lastByte) -> std::string {
      __location__("foo.spicy:16-21");
    for ( const auto& idx : contentExceptLast ) {
          __location__("foo.spicy:17-20");
        reverseOrderDigits(idx, (hilti::rt::integer::safe<std::uint8_t>{15u} & (idx >> hilti::rt::integer::safe<std::uint64_t>{4u})));
    }

      __location__("foo.spicy:22");
    return hilti::rt::fmt(std::string("%s"), lastByte);
}
```
so invoking e.g., `contextExceptLast.begin()` falls back to the iterator of the base class (`std::string` here). This is an issue similar to the one fixed in `578a4b6834f` where we only fixed `reference`, but not `iterator`. I'll have a look how much work it is change the base type to something like `std::basic_string<uchar_t>` or will just override `begin`, `end`, `cbegin` and `cend`.
